From: Ben Hutchings <ben@decadent.org.uk>
Date: Fri, 5 Jan 2018 03:09:26 +0000
Subject: x86: kvmclock: Disable use from vDSO if KPTI is enabled

Currently the pvclock pages aren't being added to user-space page
tables, and my attempt to fix this didn't work.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
 arch/x86/kernel/kvmclock.c | 5 +++++
 arch/x86/mm/kaiser.c       | 2 +-
 2 files changed, 6 insertions(+), 1 deletion(-)

--- a/arch/x86/kernel/kvmclock.c
+++ b/arch/x86/kernel/kvmclock.c
@@ -24,6 +24,7 @@
 #include <linux/percpu.h>
 #include <linux/hardirq.h>
 #include <linux/memblock.h>
+#include <linux/kaiser.h>
 
 #include <asm/x86_init.h>
 #include <asm/reboot.h>
@@ -281,6 +282,10 @@ int __init kvm_setup_vsyscall_timeinfo(v
 	if (!hv_clock)
 		return 0;
 
+	/* FIXME: Need to add pvclock pages to user-space page tables */
+	if (kaiser_enabled)
+		return 0;
+
 	size = PAGE_ALIGN(sizeof(struct pvclock_vsyscall_time_info)*NR_CPUS);
 
 	preempt_disable();
--- a/arch/x86/mm/kaiser.c
+++ b/arch/x86/mm/kaiser.c
@@ -265,7 +265,7 @@ static void __init kaiser_init_all_pgds(
  */
 void __init kaiser_init(void)
 {
-	int cpu;
+	int cpu, idx;
 
 	kaiser_init_all_pgds();
 
